File: Data Structures & Algorithms/add-two-numbers/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1==null)
        {
            return l2;
        }
        else if (l2==null)
        {
            return l1;
        }
        ListNode h1 = null;
        ListNode ans = null;
        int carry =0;
        while((l1!=null) && (l2!=null))
        {
            int chk = l1.val+l2.val+carry;
            int value = chk%10;
            carry = chk/10;
            ListNode temp = new ListNode(value);
            if(ans==null)
            {
                ans=temp;
                h1=ans;
            }
            else
            {
                ans.next = temp;
                ans = ans.next;
            }
            l1=l1.next;
            l2=l2.next;

        }
        while(l1!=null)
        {
            int chk = l1.val+carry;
            int value = chk%10;
            carry = chk/10;
            ListNode temp = new ListNode(value);
            ans.next = temp;
            ans = ans.next;
            l1=l1.next;
        }
        while(l2!=null)
        {
            int chk = l2.val+carry;
            int value = chk%10;
            carry = chk/10;
            ListNode temp = new ListNode(value);
            ans.next = temp;
            ans = ans.next;
            l2=l2.next;
        }
        if(carry!=0)
        {
            ListNode temp = new ListNode(carry);
            ans.next = temp;
        }
        return h1;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def findMid(self,head):
        slow=head
        fast=head
        while fast and fast.next:
            fast = fast.next.next
            slow =slow.next
        return slow

    def reverse(self,head):
        prev=None
        cur =head
        while cur :
            temp =cur.next
            cur.next = prev
            prev=cur
            cur =temp
        return prev

    def reorder(self,m1,m2):
        c1=m1
        c2=m2

        while c1 and c2:
            t1=c1.next
            t2=c2.next

            c1.next = c2
            c2.next = t1

            c1 =t1
            c2=t2
        
        if c2:
            c1.next=c2
        
        return m1

    def reorderList(self, head: Optional[ListNode]) -> None:
        mid =self.findMid(head)

        m1 =mid.next
        mid.next = None
        rev = self.reverse(m1)
        self.reorder(head,rev)